pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java

```diff
@@ -0,0 +1,219 @@
+        this.dataBlockFullOffset = blockSize;
+        this.entriesByteBuf = Lists.newLinkedList();
+    }
+
+    // read ledger entries.
+    private int readEntries() throws IOException {
+        checkState(bytesReadOffset >= DataBlockHeaderImpl.getDataStartOffset());
+        checkState(bytesReadOffset < dataBlockFullOffset);
+
+        try {
+            // once reach the end of entry buffer, start a new read.
+            if (entriesByteBuf.isEmpty()) {
+                readNextEntriesFromLedger();
+                log.debug("After readNextEntriesFromLedger: bytesReadOffset: {}, blockBytesHave: {}",
+                        bytesReadOffset, blockBytesHave);
+            }
+
+            // always read from the first ByteBuf in the list, once read all of its content remove it.
+            ByteBuf entryByteBuf = entriesByteBuf.get(0);
+            int ret = entryByteBuf.readByte();
+            bytesReadOffset ++;
+
+            if (entryByteBuf.readableBytes() == 0) {
+                entryByteBuf.release();
+                entriesByteBuf.remove(0);
+            }
+
+            return ret;
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Exception when get CompletableFuture<LedgerEntries>. ", e);
+            throw new IOException(e);
+        }
+    }
+
+    // read entries from ledger, and pre-handle the returned ledgerEntries.
+    private void readNextEntriesFromLedger() throws InterruptedException, ExecutionException {
+        checkState(bytesReadOffset == blockBytesHave);
+
+        long start = startEntryId + blockEntryCount;
+        long end = Math.min(start + entriesNumberEachRead - 1, ledger.getLastAddConfirmed());
+        try (LedgerEntries ledgerEntriesOnce = ledger.readAsync(start, end).get()) {
+            log.debug("read ledger entries. start: {}, end: {}", start, end);
+
+            Iterator<LedgerEntry> iterator = ledgerEntriesOnce.iterator();
+            long entryId = start;
+            while (iterator.hasNext()) {
+                LedgerEntry entry = iterator.next();
+                int entryLength = (int) entry.getLength();
+                entryId = entry.getEntryId();
+
+                if (blockSize - blockBytesHave >= entryLength + entryHeaderSize) {
+                    // data block has space for this entry, keep this entry
+                    CompositeByteBuf entryBuf = PooledByteBufAllocator.DEFAULT.compositeBuffer();
+                    ByteBuf entryHeaderBuf = PooledByteBufAllocator.DEFAULT.buffer(entryHeaderSize, entryHeaderSize);
+
+                    entryHeaderBuf.writeInt(entryLength).writeLong(entryId);
+                    entryBuf.addComponents(entryHeaderBuf, entry.getEntryBuffer().retain());
+                    entryBuf.writerIndex(entryHeaderSize + entryLength);
+
+                    entriesByteBuf.add(entryBuf);
+
+                    // set counters
+                    blockEntryCount++;
+                    payloadBytesHave += entryLength;
```

Comment:
you don't need blockBytesHave and payloadBytesHave, only payloadBytesHave.

blocksBytesHave can be calculated: payloadBytesHave + (blockEntryCount*4) + blockHeaderSize
But I don't think you need it at all.

Rename payloadBytesHave to blockEntryBytes.